internal/pkg/manifest/union.go

```diff
@@ -0,0 +1,163 @@
+// Use NewUnionA() and NewUnionB() to create a Union with an underlying
+// type already set. See Unmarshal() for details on how yaml is decoded
+// into Union.
+//
+// Union is exported to enable type embedding.
+type Union[A, B any] struct {
+	// isA is true if the underlying type of Union is A.
+	isA bool
+
+	// A holds the value of Union if IsA() is true.
+	// If IsA() is false, this is the zero value of type A.
+	//
+	// A is exported to support mergo. It should not be set
+	// directly. Use NewUnionA() to create a Union with A set.
+	A A
+
+	// isB is true if the underlying type of Union is B.
+	isB bool
+
+	// B holds the value of Union if IsB() is true.
+	// If IsB() is false, this is the zero value of type B.
+	//
+	// B is exported to support mergo. It should not be set
+	// directly. Use NewUnionB() to create a Union with B set.
+	B B
+}
+
+// NewUnionA creates a new Union[A, B] with the underlying
+// type set to A, holding val.
+func NewUnionA[A, B any](val A) Union[A, B] {
+	return Union[A, B]{
+		isA: true,
+		A:   val,
+	}
+}
+
+// NewUnionB creates a new Union[A, B] with the underlying
+// type set to B, holding val.
+func NewUnionB[A, B any](val B) Union[A, B] {
+	return Union[A, B]{
+		isB: true,
+		B:   val,
+	}
+}
+
+// IsA returns true if the underlying value of t is type A.
+func (t *Union[_, _]) IsA() bool {
+	return t.isA
+}
+
+// IsB returns true if the underlying value of t is type B.
+func (t *Union[_, _]) IsB() bool {
+	return t.isB
+}
+
+// Value returns either the underlying value of t, which is either type
+// A or B, depending on which was filled by yaml.Unmarshal or created
+// with NewUnionA/B(). If neither have been/were set, Value returns nil.
+func (t *Union[A, B]) Value() any {
+	switch {
+	case t.IsA():
+		return t.A
+	case t.IsB():
+		return t.B
+	}
+	return nil
+}
+
+// UnmarshalYAML decodes value into either type A or B, and stores that value
+// in t. value is first decoded into type A, and t will hold type A if:
+//   - There was no error decoding value into type A
+//   - A.IsZero() returns false OR A does not support IsZero()
+//
+// If A didn't meet the above criteria, then value is decoded into type B.
+// t will hold type B if B meets the same conditions that were required for type A.
+//
+// If value fails to decode into either type, or both types are zero after
+// decoding, t will not hold any value.
+func (t *Union[A, B]) UnmarshalYAML(value *yaml.Node) error {
+	// reset struct
+	var a A
+	var b B
+	t.isA, t.A = false, a
+	t.isB, t.B = false, b
+
+	aErr := value.Decode(&a)
+	if aErr == nil && !isZeroerAndZero(a) {
+		t.isA, t.A = true, a
+		return nil
+	}
+
+	bErr := value.Decode(&b)
+	if bErr == nil && !isZeroerAndZero(b) {
+		t.isB, t.B = true, b
+		return nil
+	}
+
+	if aErr != nil {
+		return aErr
+	}
+	return bErr
+}
+
+// isZeroerAndZero returns true if v is a yaml.Zeroer
+// _and_ is zero, as determined by v.IsZero().
+func isZeroerAndZero(v any) bool {
+	z, ok := v.(yaml.IsZeroer)
```

Comment:
How does this work if hc arg doesn't implement `IsZero`? Or did I miss anything 🤔  https://github.com/aws/copilot-cli/pull/4100/files#diff-4465c235880a9abf21ad9b3cab6077493fa6f767be82b60324def2bbbc3b23e2L368

Reply:
that link isn't working for me so i'm not sure what you're referencing 😢. if `v` doesn't implement `IsZero()` this func returns false. So basically when unmarshalling, if your type _does not_ implement `IsZero()`, and there was no error from `value.Decode(typeWithoutIsZero)`, then `Union` will assume it was that type.

If your type _does_ implement `IsZero()`, and there was no error from `value.Decode(typeWithZero)`, `Union` will _not_ set that type, and will try the other one or just be nil.

Reply:
An example could be:
```
type NoIsZero struct {
  Key string `yaml:"key"`
}

type HasIsZero struct {
  Key string `yaml:"key"`
}

func (z *HasIsZero) IsZero() {
  return z.Key == ""
}
```

input yaml:
```
obj:
  key: ""
```
---
Say you unmarshal into: 
```
type Obj struct {
  Union[HasIsZero, NoIsZero]
}
```
The Union's type will be NoIsZero. However, if the type was `Union[NoIsZero, HasIsZero]`, that type's Union would be NoIsZero, because there was no error unmarshalling into that type.

Does that help?

Reply:
discussed offline, changing `HTTPHealthCheckArgs.isEmpty()` to `HTTPHealthCheckArgs.IsZero()`